include/envoy/http/access_log.h

```diff
@@ -99,7 +99,7 @@ class RequestInfo {
   /**
    * @return the protocol of the request.
    */
-  virtual Protocol protocol() const PURE;
+  virtual Optional<Protocol> protocol() const PURE;
```

Comment:
For some of these things it's going to be awkward to have it all in 1 interface/class. WDYT about having a "BaseAccessLog" interface and then "HttpAccessLog", "TcpAccessLog", etc.? I don't have a good sense of what will happen but I do think it might be strange to try to mash it all together.